modules/dxp/apps/search-experiences/search-experiences-test/src/testIntegration/java/com/liferay/search/experiences/internal/upgrade/v3_1_4/test/SXPBlueprintAndSXPElementUpgradeProcessTest.java

```diff
@@ -0,0 +1,315 @@
+		for (String elementExternalReferenceCode :
+				_ELEMENT_EXTERNAL_REFERENCE_CODES) {
+
+			_createOldElement(elementExternalReferenceCode);
+		}
+
+		_runUpgrade();
+
+		for (String elementExternalReferenceCode :
+				_ELEMENT_EXTERNAL_REFERENCE_CODES) {
+
+			_assertElementUpgraded(elementExternalReferenceCode);
+		}
+	}
+
+	@Rule
+	public TestName testName = new TestName();
+
+	private void _addAssetCategories() throws Exception {
+		Company company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		_globalGroup = company.getGroup();
+
+		AssetVocabulary assetVocabulary = AssetTestUtil.addVocabulary(
+			_globalGroup.getGroupId());
+
+		_assetCategory1 = AssetTestUtil.addCategory(
+			_globalGroup.getGroupId(), assetVocabulary.getVocabularyId());
+
+		_assetCategory2 = AssetTestUtil.addCategory(
+			_globalGroup.getGroupId(), assetVocabulary.getVocabularyId());
+	}
+
+	private SXPBlueprint _addSXPBlueprint() throws Exception {
+		Class<?> clazz = getClass();
+
+		String configurationJSON = StringUtil.read(
+			clazz,
+			StringBundler.concat(
+				"dependencies/", clazz.getSimpleName(), StringPool.PERIOD,
+				testName.getMethodName(), ".configurationJSON.json"));
+
+		String elementInstancesJSON = StringUtil.read(
+			clazz,
+			StringBundler.concat(
+				"dependencies/", clazz.getSimpleName(), StringPool.PERIOD,
+				testName.getMethodName(), ".before.json"));
+
+		elementInstancesJSON = StringUtil.replace(
+			elementInstancesJSON, "$ASSET_CATEGORY_ID_1$",
+			String.valueOf(_assetCategory1.getCategoryId()));
+		elementInstancesJSON = StringUtil.replace(
+			elementInstancesJSON, "$ASSET_CATEGORY_ID_2$",
+			String.valueOf(_assetCategory2.getCategoryId()));
+		elementInstancesJSON = StringUtil.replace(
+			elementInstancesJSON, "$ASSET_CATEGORY_LABEL_1$",
+			_createAssetCategoryIDLabel(_assetCategory1));
+		elementInstancesJSON = StringUtil.replace(
+			elementInstancesJSON, "$ASSET_CATEGORY_LABEL_2$",
+			_createAssetCategoryIDLabel(_assetCategory2));
+
+		return _sxpBlueprintLocalService.addSXPBlueprint(
+			null, TestPropsValues.getUserId(), configurationJSON,
+			Collections.singletonMap(
+				LocaleUtil.US, RandomTestUtil.randomString()),
+			elementInstancesJSON, "1.1",
+			Collections.singletonMap(
+				LocaleUtil.US, RandomTestUtil.randomString()),
+			_serviceContext);
+	}
+
+	private void _assertElementUpgraded(String externalReferenceCode)
+		throws Exception {
+
+		SXPElement sxpElement =
+			_sxpElementLocalService.fetchSXPElementByExternalReferenceCode(
+				externalReferenceCode, TestPropsValues.getCompanyId());
+
+		JSONAssert.assertEquals(
+			sxpElement.getElementDefinitionJSON(),
+			_getExpectedElementDefinitionJSON(externalReferenceCode),
+			JSONCompareMode.STRICT);
+	}
+
+	private void _assertSXPBlueprint(long sxpBlueprintId) throws Exception {
+		SXPBlueprint sxpBlueprint = _sxpBlueprintLocalService.fetchSXPBlueprint(
+			sxpBlueprintId);
+
+		Assert.assertNotNull(sxpBlueprint);
```

Comment:
bchan dont like this type of assert, remove it, the test will fail if it is null.